docs_src/design/adr/device-service/0012-DeviceService-Filters.md

```diff
@@ -0,0 +1,83 @@
+- the category (such as temperature or humidity) of data as determined by the value descriptor.
+
+*With value descriptor's deprecation, what will the app function SDK offer for filtering?*
+
+### Filter by device name and value descriptor nearer collection
+
+While application service filters help to limit the amount of data sent north, those events and readings are still collected, stored, processed by the EdgeX micro services, potentially resulting in wasted resources processing unneeded data.
+
+Therefore, there is a desire to help filter data at the device service level - thereby limiting the amount of sensor data that is captured, stored and processed as well as sent north.
+
+In the first stage of device service filtering, it is desired that the filter by device name and value descriptor be offered in all device services (via the SDK).  The filtering of event/readings should be accomplished just before the event/readings are sent to core data or (or put on the message bus in future implementations where DS message application services).  See Design below for more details.
+
+### Sharing filter functions
+
+If one were to explore the filtering functions in the app functions SDK [filter.go](https://github.com/edgexfoundry/app-functions-sdk-go/blob/master/pkg/transforms/filter.go) (both FilterByDeviceName and FilterByValueDescriptor), the filters operate on a simple event model object.  Ideally, since both app services and device services share the event model (from go-mod-core-contracts), it would be the desire to share filter functions to some degree between SDKs.
+
+### Future Filters
+
+There are additional characteristics by which device services may want to filter event/readings.  Since the device service originates the event/reading, there may be circumstances where by the event/reading is considered invalid or not valuable enough to even process through the rest of EdgeX.  So, in addition to filtering by particular devices or value descriptor, device services may in the future offer filters:
+
+- based on the reading value (numeric) of a reading outside a specified range (min/max) described in the device profile for a device resource.  Thus avoiding sending in outlier or jittery data readings that could negatively effect analytics.
+- based on the reading value (numeric) equal to or near (with in some specified range) the last reading.  This allows a device service to reduce sending in event/readings that do not represent any significant change.  This filter would require that the device service maintains some sort of cache of past readings.
+
+If the concept of shared functions is achieved, there is no reason why these functions could not be reusable in application service pipelines as well.
+
+### Design Considerations
+
+#### Pipeline Nada
+
+As Device Services do not have the concept of a functions pipeline like application services do, consideration must be given as to how and where to:

```

Comment:
I think a simplified version of the functions pipeline could be implemented.

Filter function signature would be something like:
`FilterFunction(event *models.Event) *models.Event`
The event returned has been filtered. i.e. nil if at Event level or Readings collection filtered is at Readings level.

SDK would have API to set filters like` SetFilters(filters []FilterFunction)`

SDK would execute the functions the new Event in order. Execution is terminated if nil is returned resulting in the Event being filtered out. Other wise the Event returned by the last function is what is sent. 

This also allows for DS devs to create their own custom filters.

Reply:
See my reasoning for some filtering happening earlier than this stage...

Reply:
see my response on why I was thinking doing it later ;)